src/app/cluster-building-blocks/QuieterReporting.h

```diff
@@ -0,0 +1,194 @@
+#include <app/data-model/Nullable.h>
+#include <system/SystemClock.h>
+
+namespace chip {
+namespace app {
+
+// - If it has changed due to a change in the CurrentPhase or OperationalState attributes, or
+// - When it changes from 0 to any other value and vice versa, or
+// - When it changes from null to any other value and vice versa, or
+// - When it increases, or
+// - When there is any increase or decrease in the estimated time remaining that was due to progressing insight of the server's control logic, or
+// - When it changes at a rate significantly different from one unit per second.
+
+enum class QuieterReportingPolicyEnum
+{
+    kMarkDirtyOnChangeToFromZero = (1u << 0),
+    kMarkDirtyOnDecrement = (1u << 1),
+    kMarkDirtyOnIncrement = (1u << 2),
+};
+
+using QuieterReportingPolicyFlags = ::chip::BitFlags<QuieterReportingPolicyEnum>;
+
+namespace detail {
+
+using Timestamp = chip::System::Clock::Milliseconds64;
+template <typename T>
+using Nullable = chip::app::DataModel::Nullable<T>;
+
+/**
+ * This class helps track reporting state of an attribute to properly keep track of whether
+ * it needs to be marked as dirty or not for purposes of reporting using Q quality.
+ *
+ * The class can be configured via `SetPolicy` to have some/all of the common reasons
+ * for reporting (e.g. increment only, decrement only, change to/from zero).
+ *
+ * Changes of null to non-null or non-null to null are always considered dirty.
+ *
+ * It is possible to force mark the attribute as dirty (see `ForceMarkAsDirty`) such as
+ * for conditions like "When there is any increase or decrease in the estimated time
+ * remaining that was due to progressing insight of the server's control logic".
+ *
+ * Usage is simple:
+ *
+ * - Call `SetValue()` with the new value and current monotonic system timestamp
+ *   - There is an overload with a `SufficientChangePredicate` which will apply externally
+ *     provided checks between old/new value and time last marked dirty to allow for complex
+ *     rules like marking dirty less than once per second, etc.
+ * - *Maybe* call `ForceMarkAsDirty()` in some choice situations (e.g. know for sure it's
+ *    an important update, like at the edge of an operational state change).
+ * - Call `GetThenResetDirtyState()`. If it returns true, mark the attribute dirty, with the
+ *   method most suitable at the call site (e.g. `MatterReportingAttributeChangeCallback` call
+ *   or similar methods).
+ *
+ * @tparam T - the type of underlying numerical value that will be held by the class.
+ */
+template <typename T>
+class QuieterReportingAttribute
+{
+  public:
+    explicit QuieterReportingAttribute(const chip::app::DataModel::Nullable<T>& initialValue) : mValue(initialValue), mLastDirtyValue(initialValue) {}
+
+    using SufficientChangePredicate = std::function<bool(Timestamp /* previousDirtyTime */, Timestamp /* now */, const Nullable<T> & /* previousDirtyValue */, const Nullable<T> & /* newValue */)>;
```

Comment:
How about passing a struct here instead of 4 arguments? I am thinking that would get us:
   - smaller signature (4 arguments start being large)
   - nicer backwards compatibility in the unlikely case we want to pass in more things here
   - enforced naming (otherwise both callers and implementors may have a much easier time to mis-order same-type of values ... we have 2 identical timestamps and nullable<t> so switching order is very easy)

We need some documentation: what does previousDirtyTime mean? It seems that if the previous value was not dirty we default-init the milliseconds there as {}. Is that 0 or current time? Would a optional not be more correct here?

Reply:
Will rework.

Reply:
Done